### DISCUSSION TITLE
Necessity of overriding display-buffer-alist

### DISCUSSION THREAD

**Discussion by ktfleming** [bard/emacs-run-command#19] [Q&A]
Hi! Thanks for this great package that I use dozens of times per day -- it's become an invaluable part of my workflow.

I'm curious about the let-binding on `display-buffer-alist` here: https://github.com/bard/emacs-run-command/blob/43afe908be26f13f5801dff69179aa64e74a235f/run-command-core.el#L74-L75

Looking at the history it was introduced in https://github.com/bard/emacs-run-command/commit/e9420f1527dff8a65aeb36f8c0605ce8ad7cfb04, but I'm not sure how/if it's necessary to achieve the feature `when command is already running, switch to it` (from the commit message). Do you think this binding is indeed necessary?

I ask because I was trying to use `display-buffer-alist` to control how the buffers created by run-command appear, and found that it wasn't working, which led me to this. Removing it did get me the behavior that I wanted, and I haven't come across any odd side-effects yet, but I'm wondering if I might encounter some. Or if it's not actually necessary, maybe this binding could be removed in run-command.

**Comment by bard** [comment]
> Hi! Thanks for this great package that I use dozens of times per day -- it's become an invaluable part of my workflow.

I'm really glad to hear!

> I ask because I was trying to use `display-buffer-alist` to control how the buffers created by run-command appear, and found that it wasn't working, which led me to this. Removing it did get me the behavior that I wanted, and I haven't come across any odd side-effects yet, but I'm wondering if I might encounter some. Or if it's not actually necessary, maybe this binding could be removed in run-command.

I added it (begrudgingly) because I was hitting situations where the commands would show up in the current rather than the "other" window. I should at least have added a comment because I can't off the top of my head remember the exact sequence of steps. But yes, it's a heavy-handed solution. I'll try to reproduce the issue and post back here, if you're using `display-buffer-alist` routinely you might know of a better way.

**Reply by ktfleming** [reply]
I see -- in my case I ended up doing this:

```elisp
(defun run-command-buffer-p (buffer arg)
  "Return non-nil if BUFFER is a buffer for a `run-command' invocation."
  (with-current-buffer buffer
    ;; TODO: is there a better way to determine this?
    (derived-mode-p 'term-mode)))

(setq display-buffer-alist
      '((run-command-buffer-p
         (display-buffer-reuse-mode-window display-buffer-pop-up-window))))

```

This seems to get me the behavior I was looking for, which is that command invocations will reuse any existing window that's already displaying a `term-mode` buffer, or creating a new window if one doesn't exist. But it only works for me because I only use `term-mode` for run-command commands -- if I used "regular" `term-mode` buffers, this would probably be too general.

I wonder if one solution would be to have a way of determining whether a given buffer is one created by run-command and not some other term/vterm/compilation buffer, i.e. filling in that TODO above; that way all users could customize their own `display-buffer-alist` to have the commands appear how they like (me using what I have above, perhaps you use `display-buffer-use-least-recent-window`, etc).

**Comment by bard** [comment]
You could use:

```emacs-lisp
(defun run-command-buffer-p (buffer arg)
  (boundp 'run-command--command-spec))
```
`run-command--command-spec` is internal, so no guarantees of stability, but I think having a function to check whether a buffer is `run-command` output is a good idea, I'll probably add one like the above. 

Is the second parameter (`arg`) necessary?

> I wonder if one solution would be to have a way of determining whether a given buffer is one created by run-command and not some other term/vterm/compilation buffer, i.e. filling in that TODO above; that way all users could customize their own `display-buffer-alist` to have the commands appear how they like (me using what I have above, perhaps you use `display-buffer-use-least-recent-window`, etc).

Customizability is important but a reliable default behavior has higher priority — what to do will depend on those scenarios I mentioned above (haven't got around to trying to reproduce them yet).

**Reply by ktfleming** [reply]
> Is the second parameter (`arg`) necessary?

Good question, I only had it there because from what I understand using a function in this way in `display-buffer-alist` will pass it to `buffer-match-p`, whose docstring says that it accepts `a predicate function that takes a buffer object and ARG as
  arguments` -- but I tried removing it and the behavior seems the same (`arg` is always nil in my testing) so it looks like it's probably not necessary.

> Customizability is important but a reliable default behavior has higher priority — what to do will depend on those scenarios I mentioned above (haven't got around to trying to reproduce them yet).

Yep, that sounds good.

**Reply by bard** [reply]
@ktfleming if I comment out the `display-buffer-alist` override, this happens on my system:

1. begin with a single-window layout
2. execute `M-x run-command` and select any command
   - frame splits (in two side-by-side windows in my case)
   - command output appears in secondary window
   - focus stays in current window
3. execute `M-x delete-window`
   - frame contains single window
   - single window contains command buffer
4. execute `M-x switch-to-buffer` and select initial buffer
   - frame contains a single window
   - single window contains initial buffer
5. execute `M-x split-window-right`
   - frame contains two windows
   - both windows contain initial buffer
6. execute `M-x run-command` and select previously selected command
   - **command output should appear in other window, but it appears in current window**

Does that happen on your system too?

**Reply by ktfleming** [reply]
@bard I'm not seeing the same behavior, starting in step 3. After doing `delete-other-window` (actually I don't have that command but I'm assuming you meant `delete-other-windows`?), I end up with a single window, but it's showing the initial buffer, not the command buffer. Since in step 2 you end with the selected window still showing the initial buffer, it seems strange that doing `delete-other-windows` would leave you with only the command buffer showing, rather than the initial buffer.

Ignoring that and skipping step 4 (since I already have the initial buffer in the frame's sole window), following steps 5 and 6 results in the command output appearing in the other (non-selected) window, rather than in the current window like you're seeing.

I'm not sure what the reason for the discrepancy could be -- my testing was with the `display-buffer-alist` override removed, and my own `display-buffer-alist` set to `nil`.

**Reply by bard** [reply]
Sorry @ktfleming, the sequence contained a mistake, `delete-other-window` was meant to be `delete-window`. Could you have another try following the edit?

**Reply by ktfleming** [reply]
@bard Sorry for the delay, this slipped my mind. With the updated instructions and without the `display-buffer-alist` override I do see the behavior you describe, and agree that it's not ideal.

**Reply by bard** [reply]
No worries!

Can you check whether this works for you?

```diff
@@ -72,9 +72,17 @@ See `run-command-recipes' for the expected format of COMMAND-SPEC."
           (when (map-contains-key command-spec :hook)
             (funcall (map-elt command-spec :hook))))))
     (let ((display-buffer-alist
-           '((".*" display-buffer-use-least-recent-window))))
+           (append
+            display-buffer-alist
+            '((run-command-buuffer-p display-buffer-use-least-recent-window)))))
       (display-buffer buffer))))
 
+(defun run-command-buffer-p (buffer &optional _arg)
+  "Tell whether the given BUFFER was created by `run-command'.
+
+`ARG' is unused and only defined due to be required by `display-buffer-alist'."
+  (boundp 'run-command--command-spec))
+
 ;;;; Internals
 
 (defvar-local run-command--command-spec nil
```

The idea is to temporarily add the fallback rule at the end of the `display-buffer-alist`, so if you previously added your own rule, it will take precedence.

If it works, I'll make a release with it soon.

**Reply by ktfleming** [reply]
That's working for me, in doing some testing it seems my custom `display-buffer-alist` rule is getting prioritized over the fallback. 👍 
(You might have noticed but there's a typo in the patch: `run-command-buuffer-p`. Just pointing it out to prevent the possibility of having to make a bugfix release 🙂 )

**Reply by bard** [reply]
> (You might have noticed but there's a typo in the patch: `run-command-buuffer-p`. Just pointing it out to prevent the possibility of having to make a bugfix release slightly_smiling_face )

Great catch. +1

I just pushed this and the vterm fix,  MELPA should pick those up shortly.

Thank you for reporting and for helping to test and debug this!

**Comment by vkazanov** [comment]
Hi @bard ,

Are you sure tweaking display-buffer-alist is the right option after all? I had to patch it out in the latest stable version using =el-patch= to fix the behaviour.

Quoting the [Zen of buffer display](https://www.gnu.org/software/emacs/manual/html_node/elisp/The-Zen-of-Buffer-Display.html):

> display-buffer-alist and display-buffer-base-action are user options—Lisp programs must never set or rebind them. 

Great package, btw, use it hourly!

Thanks
